### PR TITLE
[Merged by Bors] - feat(RingTheory/Smooth): Correspondence between square-zero lifts and retractions of the conormal sequence.

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Smooth/Projective.lean [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+lemma sectionOfRetractionKerToTensorAux_algebraMap (x : P) :
+    sectionOfRetractionKerToTensorAux hf' l hl σ hσ (algebraMap P S x) = x - l (1 ⊗ₜ .D _ _ x) :=
+  sectionOfRetractionKerToTensorAux_prop l hl _ x (by simp [hσ])
+
+/--
+Given a surjective algebra hom `f : P →ₐ[R] S` with square-zero kernel `I`,
+Suppose we have a retraction `l` of the injection `l : I →ₗ[P] S ⊗[P] Ω[P/R]`,
+then `x ↦ σ x - l (1 ⊗ D x)` is an algebra homomorphism and a section to `f`,
+where `σ` is an arbitrary (set-theoretic) section of `f`
```
Comment (chrisflav) [inline Mathlib/RingTheory/Smooth/Projective.lean:157]: same as above

File: Mathlib/RingTheory/Smooth/Projective.lean [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    sectionOfRetractionKerToTensor hf hf' l hl (algebraMap P S x) = x - l (1 ⊗ₜ .D _ _ x) :=
+  sectionOfRetractionKerToTensorAux_algebraMap hf' l hl _ _ x
+
+@[simp]
+lemma toAlgHom_comp_sectionOfRetractionKerToTensor :
+    (IsScalarTower.toAlgHom R P S).comp
+      (sectionOfRetractionKerToTensor hf hf' l hl) = AlgHom.id _ _ := by
+  ext x
+  obtain ⟨x, rfl⟩ := hf x
+  simp [(RingHom.mem_ker _).mp]
```
Comment (chrisflav) [inline Mathlib/RingTheory/Smooth/Projective.lean:173]: Maybe show this for `sectionOfRetractionKerToTensorAux` and then specialize?

File: Mathlib/RingTheory/Smooth/Projective.lean [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+end ofRetraction
+
+/--
+Given a surjective algebra hom `f : P →ₐ[R] S` with square-zero kernel `I`,
+there is a one-to-one correspondance between algebra homomorphism sections of `f`,
+and `P`-linear retractions to `I →ₗ[P] S ⊗[P] Ω[P/R]`.
```
Comment (chrisflav) [inline Mathlib/RingTheory/Smooth/Projective.lean:180]: The equiv is defined the other way round.

File: Mathlib/RingTheory/Smooth/Projective.lean [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+-- Suppose we have a section (as alghom) of `P →ₐ[R] S`.
+variable (g : S →ₐ[R] P) (hg : (IsScalarTower.toAlgHom R P S).comp g = AlgHom.id R S)
+
+/--
+Given a surjective algebra hom `f : P →ₐ[R] S` with square-zero kernel `I`,
+and a section `g : S →ₐ[R] P` (as an algebra hom),
+we get a `R`-derivation `P → I` via `x ↦ x - g (f x)`.
+-/
+@[simps]
+def derivationOfSectionOfKerSqZero : Derivation R P (RingHom.ker (algebraMap P S)) where
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/RingTheory/Smooth/Projective.lean]: It is morally in the assumption as implied by the existence of a section.

File: Mathlib/RingTheory/Ideal/Cotangent.lean [hunk 5]
```diff
@@ -114,6 +114,12 @@ theorem toCotangent_to_quotient_square (x : I) :
     I.cotangentToQuotientSquare (I.toCotangent x) = (I ^ 2).mkQ x := rfl
 #align ideal.to_cotangent_to_quotient_square Ideal.toCotangent_to_quotient_square
 
+lemma Cotangent.smul_eq_zero_of_mem {R} [CommRing R] {I : Ideal R}
+    {x} (hx : x ∈ I) (m : I.Cotangent) : x • m = 0 := by
+  obtain ⟨m, rfl⟩ := Ideal.toCotangent_surjective _ m
+  rw [← map_smul, Ideal.toCotangent_eq_zero, pow_two]
+  exact Ideal.mul_mem_mul hx m.2
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/RingTheory/Ideal/Cotangent.lean]: I added it as a separate lemma.

**Review by chrisflav — APPROVED** [review]
Sorry, I forgot about this PR. Besides some (very) minor typos, I think this is good to go.

File: Mathlib/RingTheory/Smooth/Kaehler.lean [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+end ofRetraction
+
+/--
+Given a surjective algebra homomorphism `f : P →ₐ[R] S` with square-zero kernel `I`,
+there is a one-to-one correspondence between `P`-linear retractions of `I →ₗ[P] S ⊗[P] Ω[P/R]`
+and algebra homomorphism sections of `f`.
+-/
+noncomputable
+def retractionEquivSectionKerToTensor :
```
Comment (jcommelin) [inline Mathlib/RingTheory/Smooth/Kaehler.lean:193]: Is this name the wrong way round? `kerToTensor` is on the LHS...

**Review by jcommelin — COMMENTED** [review]
This looks good to me. (1 minor comment)

@riccardobrasca do you want to have another look?

**Comment by riccardobrasca** [comment]
Sorry for the late answer, LGTM, thanks!

bors merge

**Pull request closed without merging**